public/app/plugins/panel/logs/LogsPanel.tsx

```diff
@@ -37,14 +44,36 @@ export const LogsPanel = ({
     sortOrder,
     dedupStrategy,
     enableLogDetails,
+    showLogContextToggle,
   },
-  title,
   id,
 }: LogsPanelProps) => {
   const isAscending = sortOrder === LogsSortOrder.Ascending;
   const style = useStyles2(getStyles);
   const [scrollTop, setScrollTop] = useState(0);
   const logsContainerRef = useRef<HTMLDivElement>(null);
+  const [contextRow, setContextRow] = useState<LogRowModel | null>(null);
+  const [closeCallback, setCloseCallback] = useState<(() => void) | null>(null);
+
+  // create a map of refid -> datasource
```

Comment:
I don't know how you feel about this, but when components start growing in complexity and responsibilities it feels particularly uncomfortable, as it becomes challenging to maintain and test.

Reply:
100% agree. I extracted that logic into an additional hook. WDYT?